Add lerping between LinearGradients with arbitrary number of colors and stops

spkersten opened the pull request.

goderbauer commented:
Thanks for the contribution.

This is slightly changing the behavior for the existing functionality of lerping between Gradients with equal stops. Previously, the stops would gradually move from their location in Gradient `a` to their new location in Gradient `b` as `t` is increased. With this change, the stops don't move. Instead, the Gradients appear to cross-fade from one to another. I captured the differences in the video below (watch the yellow line). This is not necessarily a bad change, but I was wondering if this was your intention?

Old behavior:
![ezgif-4-5a4e84417129](https://user-images.githubusercontent.com/1227763/52384154-1b391280-2a31-11e9-8a1a-ea77e72472e3.gif)

New behavior:
![ezgif-4-5484546c2358](https://user-images.githubusercontent.com/1227763/52384193-491e5700-2a31-11e9-8810-e0e0c712d7b5.gif)

Code used to generate:

```dart
import 'package:flutter/material.dart';

class AppBarBug extends StatefulWidget {
  @override
  _AppBarBugState createState() => new _AppBarBugState();
}

class _AppBarBugState extends State<AppBarBug> with SingleTickerProviderStateMixin {

  AnimationController _controller;

  @override
  void initState() {
    super.initState();
    _controller = AnimationController(vsync: this, duration: const Duration(seconds: 5));
    _controller.repeat(reverse: true);
  }

  @override
  Widget build(BuildContext context) {
    LinearGradient a = LinearGradient(
      stops: [0.0, 0.8, 1.0],
      colors: [
        const Color(0xFFFFFFEE),
        Colors.green,
        const Color(0xFF999999),
      ],
      tileMode: TileMode.repeated,
    );
    LinearGradient b = LinearGradient(
      stops: [0.0, 0.2, 1.0],
      colors: [
        Colors.red,
        Colors.yellow,
        Colors.blue,
      ],
      tileMode: TileMode.repeated,
    );

    return Scaffold(
      body: Center(
        child: AnimatedBuilder(
          animation: _controller,
          builder: (BuildContext context, _) => Container(
            height: 100.0,
            width: 200.0,
            decoration: BoxDecoration(
              gradient: LinearGradient.lerp(a, b, _controller.value),
            ),
          ),
        ),
      ),
    );
  }
}

void main() {
  runApp(new MaterialApp(home: new AppBarBug()));
}
```

goderbauer commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -374,16 +374,34 @@ class LinearGradient extends Gradient {
       return b.scale(t);
     if (b == null)
       return a.scale(1.0 - t);
-    final _ColorsAndStops interpolated = _interpolateColorsAndStops(a.colors, a.stops, b.colors, b.stops, t);
```
Any reason you can't add the new logic to the `_interpolateColorsAndStops`, so the other Gradient classes can also profit from this behavior?

goderbauer replied:
(There's a TODO in that function for exactly this functionality)

spkersten replied:
I've moved the logic there

goderbauer commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -374,16 +374,34 @@ class LinearGradient extends Gradient {
     if (b == null)
       return a.scale(1.0 - t);
-    final _ColorsAndStops interpolated = _interpolateColorsAndStops(a.colors, a.stops, b.colors, b.stops, t);
+    final List<double> combinedStops = ((a._impliedStops() ?? <double>[0, 1]) + (b._impliedStops() ?? <double>[0, 1]))..sort();
```
These 5 lines could probably be simplified by adding `(a._impliedStops() ?? <double>[0, 1])` and `(b._impliedStops() ?? <double>[0, 1])` to a `SplayTreeSet` [1]?

[1] https://master-docs.flutter.io/flutter/dart-collection/SplayTreeSet-class.html

goderbauer commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -374,16 +374,34 @@ class LinearGradient extends Gradient {
+      if (!uniqueCombinedStops.contains(stop))
+        uniqueCombinedStops.add(stop);
+    final List<Color> interpolatedColors = uniqueCombinedStops.map<Color>(
+            (double s) => Color.lerp(a._sample(s), b._sample(s), t)).toList(growable: false);
```
nit: formatting here should be

```dart
    final List<Color> interpolatedColors = uniqueCombinedStops.map<Color>(
      (double s) => Color.lerp(a._sample(s), b._sample(s), t)
    ).toList(growable: false);
```

goderbauer commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -374,16 +374,34 @@ class LinearGradient extends Gradient {
+    if (t >= stops.last)
+      return colors.last;
+    final int index = stops.lastIndexWhere((double s) => s <= t);
+    return Color.lerp(
```
nit: formatting here should be:

```dart
  return Color.lerp(
    colors[index], colors[index + 1],
    (t - stops[index]) / (stops[index + 1] - stops[index])
  );
```

goderbauer commented on packages/flutter/test/painting/gradient_test.dart:
```diff
@@ -91,11 +92,13 @@ void main() {
       end: Alignment(-1.0, 0.0),
       colors: <Color>[
         Color(0x3B3B3B3B),
+        Color(0x55555555),
```
Instead of editing an existing test case please add new tests specific to this change and keep the old test cases as-is to ensure that those cases keep working as well with your change.

spkersten replied:
Maybe I'm misunderstanding you, but I changed the behaviour so the test is still valid but the expected output has changed.

goderbauer replied:
Oh, sorry. I wasn't looking at the full context here.

spkersten commented:
@goderbauer The behaviour change was indeed intended (and hopefully desired). I don't see how it would be possible to keep the old behaviour when the number of stops doesn't match. A point-wise interpolation as in this PR seems to give a more predictable result.

goderbauer commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -16,22 +17,44 @@ class _ColorsAndStops {
+  aStops ??= _impliedStops(aColors.length);
+  bStops ??= _impliedStops(bColors.length);
+  final SplayTreeSet<double> stops = SplayTreeSet<double>(
+          (double a, double b) => (a - b).sign.round())
```
Why's the default comparator not sufficient? Seems like it just uses double.compareTo, which should be fine, no?

goderbauer commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -16,22 +17,44 @@ class _ColorsAndStops {
+  final double separation = 1.0 / (number - 1);
+  return List<double>.generate(
+    number,
+        (int index) => index * separation,
```
nit: remove extra whitespace.

```suggestion
    (int index) => index * separation,
```

goderbauer commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -16,22 +17,44 @@ class _ColorsAndStops {
+}
+
+_ColorsAndStops _interpolateColorsAndStops(
+    List<Color> aColors, List<double> aStops, List<Color> bColors, List<double> bStops, double t) {
```
nit: add some line breaks?

```suggestion
    List<Color> aColors, 
    List<double> aStops, 
    List<Color> bColors,
    List<double> bStops, 
    double t,
) {
```

goderbauer commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -16,22 +17,44 @@ class _ColorsAndStops {
+          (double a, double b) => (a - b).sign.round())
+    ..addAll(aStops)
+    ..addAll(bStops);
+  final List<double> interpolatedStops = stops.toSet().toList()..sort();
```
Why the `toSet` and `sort` calls? I'd expect that `stops.toList()` already gives you a properly sorted list since stops is a SplayTreeSet?

goderbauer replied:
Also, the `toList` call could probably specify `growable: false`?

goderbauer commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -16,22 +17,44 @@ class _ColorsAndStops {
+    ..addAll(bStops);
+  final List<double> interpolatedStops = stops.toSet().toList()..sort();
+  final List<Color> interpolatedColors = interpolatedStops.map<Color>(
+          (double s) => Color.lerp(_sample(aColors, aStops, s), _sample(bColors, bStops, s), t)
```
Can you rename `s` to `stop` for clarity?

goderbauer commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -16,22 +17,44 @@ class _ColorsAndStops {
+  final int index = stops.lastIndexWhere((double s) => s <= t);
+  return Color.lerp(
+      colors[index], colors[index + 1],
+      (t - stops[index])/(stops[index + 1] - stops[index])
```
nit: spaces around `/`

goderbauer commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -16,22 +17,44 @@ class _ColorsAndStops {
-    for (int i = 0; i < aStops.length; i += 1)
-      interpolatedStops.add(ui.lerpDouble(aStops[i], bStops[i], t).clamp(0.0, 1.0));
-  }
+List<double> _impliedStops(int number) {
```
Can you keep this as a private function within Gradient and just change the call sites of _interpolateColorsAndStops to pass in the result of a call to this function for the stops parameters? (You'll have to update the assert in _interpolateColorsAndStops).

goderbauer replied:
(if it's pulled out like this it's kinda unclear what the `number` parameter is. It was cleared before where this would just access a member.

goderbauer commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -335,14 +344,14 @@ class LinearGradient extends Gradient {
 
   @override
   Gradient lerpFrom(Gradient a, double t) {
-    if (a == null || (a is LinearGradient && a.colors.length == colors.length)) // TODO(ianh): remove limitation
```
This todo also exists on the other Gradient subclasses in this file. I believe they can be removed there as well, no?

goderbauer commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -16,22 +17,44 @@ class _ColorsAndStops {
+  );
+}
+
+Color _sample(List<Color> colors, List<double> stops, double t) {
```
What is t? Can you give that a more descriptive name?

spkersten replied:
I struggle to find a name for this. It seems the same as `t` in, for example, the `lerp` methods.

goderbauer replied:
In that case: Can you add a short doc comment to this method explaining what it does for the future reader, something like "Calculate color at position `t` of the gradient defined by `colors` and `stops`."

goderbauer reviewed (commented):
This will be a great feature to have!

goderbauer commented:
You'll have to add some new tests that exercise the newly added functionality, though. E.g. some tests that lerp between an arbitrary number of colors and stops in addition to the excising tests where colors.length == stops.length.

spkersten committed:
Add lerping between Gradients with arbitrary number of colors and stops

spkersten commented:
@goderbauer I should have addressed all your comments. Let me know if the test coverage is enough.(I'm happy to add more, but I also don't want to increase maintenance by adding too many tests)

goderbauer commented:
Thanks. Before I take an in-deeps look: Can you please fix the whitespace errors the analyzer check is complaining about so the tests can run properly.

All these lines have trailing whitespace, which should be removed:

```
packages/flutter/lib/src/painting/gradient.dart:33:    List<Color> aColors, 
packages/flutter/lib/src/painting/gradient.dart:34:    List<double> aStops, 
packages/flutter/lib/src/painting/gradient.dart:36:    List<double> bStops, 
```

goderbauer commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -16,22 +17,37 @@ class _ColorsAndStops {
+  assert(bColors.length >= 2);
+  assert(aStops.length == aColors.length);
+  assert(bStops.length == bColors.length);
+  final SplayTreeSet<double> stops = SplayTreeSet<double>(
```
Isn't comareTo the default if nothing is provided? Can these lines be changed to:

```dart
final SplayTreeSet<double> stops = SplayTreeSet<double>()
    ..addAll(aStops)
    ..addAll(bStops);
```

spkersten replied:
That makes sense. Strange, somehow I missed that completely.

goderbauer commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -873,7 +899,7 @@ class SweepGradient extends Gradient {
     if (b == null)
       return a.scale(1.0 - t);
-    final _ColorsAndStops interpolated = _interpolateColorsAndStops(a.colors, a.stops, b.colors, b.stops, t);
+    final _ColorsAndStops interpolated = _interpolateColorsAndStops(a.colors, a._impliedStops(), b.colors, b._impliedStops(), t);
```
nit for consistency: can you format this line in the same way everywhere?

goderbauer reviewed (approved):
LGTM modulo three nits.

Thank you for this contribution!

spkersten commented:
Fixed hopefully the last mistakes. Thanks for your patience @goderbauer

dnfield commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -16,22 +17,37 @@ class _ColorsAndStops {
-  }
+/// Calculate the color at position [t] of the gradient defined by [colors] and [stops].
+Color _sample(List<Color> colors, List<double> stops, double t) {
+  if (t <= stops.first)
```
Can we get some asserts here that `stops != null` and `stops.isNotEmpty`?

dnfield replied:
We should also `assert(t != null)`

dnfield commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -16,22 +17,37 @@ class _ColorsAndStops {
+/// Calculate the color at position [t] of the gradient defined by [colors] and [stops].
+Color _sample(List<Color> colors, List<double> stops, double t) {
+  if (t <= stops.first)
+    return colors.first;
```
Same asserts for `colors`

dnfield commented on packages/flutter/lib/src/painting/gradient.dart:
```diff
@@ -16,22 +17,37 @@ class _ColorsAndStops {
+    return colors.first;
+  if (t >= stops.last)
+    return colors.last;
+  final int index = stops.lastIndexWhere((double s) => s <= t);
```
We should check for error conditions here (`if (index == -1) ...`)

dnfield replied:
Should be fine here to just `assert(index != -1);` based on other variants of this method

spkersten replied:
Added the assert since getting `-1` should be impossible because there is at least one stop `<= t` at this point.

dnfield reviewed (changes requested):
This LGTM once the asserts/error check are in place.  I know it's a private method and we're asserting things before it, but it will make this harder to break if someone refactors or adds another call path later.
